Add emcee as a possible sampler to pycbc inference

cdcapano opened the pull request:
This patch adds emcee's EnsembleSampler to the list of possible samplers in pycbc inference. Since emcee has no burn-in feature, `pycbc_inference` must be run with `--skip-burn-in` if using emcee. This patch also adds `emcee` to the list of required packages in `setup.py`. I tested this on my laptop with 4 processes and it seems to work well.

ahnitz commented on pycbc/inference/sampler.py:
```diff
@@ -64,14 +64,45 @@ def chain(self):
         """ This function should burn in the sampler.
         """
-        raise ValueError("burn_in function not set.")
+        raise ValueError("This sampler has no burn_in function.")
```
Use NotImplementedErrror

cdcapano commented:
There was a bug that I just fixed: kombine returns a chain that is the transpose of what emcee returns. This is fixed now. Also, I fixed the things landscape was complaining about.

cmbiwer commented on pycbc/inference/sampler.py:
```diff
@@ -181,8 +276,14 @@ def run(self, niterations, **kwargs):
+
 
 samplers = {
     "kombine" : KombineSampler,
```
Both kombine and emcee are ensemble samplers. I think should call them either ``KombineEnsembleSampler`` and ``EmceeEnsembleSampler`` or ``KombineSampler`` or ``EmceeSampler``.

cmbiwer commented on pycbc/inference/sampler.py:
```diff
@@ -181,8 +276,14 @@ def run(self, niterations, **kwargs):
         """
-        return self._sampler.run_mcmc(niterations, **kwargs)
+        try:
+            p0 = kwargs.pop('p0')
```
If ``p0`` is required why not put it in the docstring and as an option? This kinds of hides that its required.

cdcapano commented:
@cmbiwer I addressed your comments. For `p0`, I changed it so that the Sampler class keeps track of the position and initial variables. Now you just call burn_in and run_mcmc without needing to pass it a `p0` for both the kombine case and the emcee case.

For the naming of `EnsembleSampler`: I called it that just because emcee has 3 different samplers you can use, one of which is called `EnsembleSampler`; see:
http://dan.iel.fm/emcee/current/api/
We'll probably never want to use the Metropolis-Hastings sampler, but we may want to add support for the parallel tempered one at some point. In that case we'll want some naming convention to define the two.

One other thing that's changed: it turns out that kombine and emcee define acceptance fraction differently. Kombine returns the number of walkers that accepted the proposal at each iteration (thus returning an niterations-long array) while emcee returns the number of iterations accepted by each walker (thus returning an nwalkers-long array). Kombine can also return this later value, but it's called `acceptance_rate` there. This was causing the code to crash when trying to write the acceptance fraction to file. I'm not 100% sure what to do here, but for the moment, I just added a catch in `pycbc_inference` to make the acceptance fraction array the appropriate size depending on if kombine or emcee is used. It would be good to fix this at some point in the sampler module, but I'd rather not hold up the entire patch just for this at the moment.
